// testcafe/lib/notifications/warning-log.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const render_template_1 = __importDefault(require("../utils/render-template"));
const message_bus_1 = __importDefault(require("../utils/message-bus"));
class WarningLog {
    constructor(globalLog = null, callback) {
        this.globalLog = globalLog;
        this.messages = [];
        this.callback = callback;
    }
    addPlainMessage(msg) {
        // NOTE: avoid duplicates
        if (!this.messages.includes(msg))
            this.messages.push(msg);
    }
    addWarning(...args) {
        // @ts-ignore
        const msg = render_template_1.default.apply(null, args);
        this.addPlainMessage(msg);
        if (this.globalLog)
            this.globalLog.addPlainMessage(msg);
        if (this.callback)
            this.callback(msg);
    }
    clear() {
        this.messages = [];
    }
    copyTo(warningLog) {
        this.messages.forEach(msg => warningLog.addWarning(msg));
    }
    static createAddWarningCallback(messageBus, testRun) {
        return async (message) => {
            if (messageBus && messageBus instanceof message_bus_1.default) {
                await messageBus.emit('warning-add', {
                    message,
                    testRun,
                });
            }
        };
    }
}
exports.default = WarningLog;
module.exports = exports.default;
